pkg/logql/filter_test.go

```diff
@@ -0,0 +1,141 @@
+	}{
+		// regex we intend to support.
+		{"foo", true, literalFilter([]byte("foo")), true},
+		{"not", true, newNotFilter(literalFilter([]byte("not"))), false},
+		{"(foo)", true, literalFilter([]byte("foo")), true},
+		{"(foo|ba)", true, newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("ba"))), true},
+		{"(foo|ba|ar)", true, newOrFilter(newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("ba"))), literalFilter([]byte("ar"))), true},
+		{"(foo|(ba|ar))", true, newOrFilter(literalFilter([]byte("foo")), newOrFilter(literalFilter([]byte("ba")), literalFilter([]byte("ar")))), true},
+		{"foo.*", true, literalFilter([]byte("foo")), true},
+		{".*foo", true, newNotFilter(literalFilter([]byte("foo"))), false},
+		{".*foo.*", true, literalFilter([]byte("foo")), true},
+		{"(.*)(foo).*", true, literalFilter([]byte("foo")), true},
+		{"(foo.*|.*ba)", true, newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("ba"))), true},
+		{"(foo.*|.*bar.*)", true, newNotFilter(newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("bar")))), false},
+		{".*foo.*|bar", true, newNotFilter(newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("bar")))), false},
+		{".*foo|bar", true, newNotFilter(newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("bar")))), false},
+		// This construct is similar to (...), but won't create a capture group.
+		{"(?:.*foo.*|bar)", true, newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("bar"))), true},
+		// named capture group
+		{"(?P<foo>.*foo.*|bar)", true, newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("bar"))), true},
+		// parsed as (?-s:.)*foo(?-s:.)*|b(?:ar|uzz)
+		{".*foo.*|bar|buzz", true, newOrFilter(literalFilter([]byte("foo")), newOrFilter(literalFilter([]byte("bar")), literalFilter([]byte("buzz")))), true},
+		// parsed as (?-s:.)*foo(?-s:.)*|bar|uzz
+		{".*foo.*|bar|uzz", true, newOrFilter(newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("bar"))), literalFilter([]byte("uzz"))), true},
+		// parsed as foo|b(?:ar|(?:)|uzz)|zz
+		{"foo|bar|b|buzz|zz", true, newOrFilter(newOrFilter(literalFilter([]byte("foo")), newOrFilter(newOrFilter(literalFilter([]byte("bar")), literalFilter([]byte("b"))), literalFilter([]byte("buzz")))), literalFilter([]byte("zz"))), true},
+		// parsed as f(?:(?:)|oo(?:(?:)|bar))
+		{"f|foo|foobar", true, newOrFilter(literalFilter([]byte("f")), newOrFilter(literalFilter([]byte("foo")), literalFilter([]byte("foobar")))), true},
+		// parsed as f(?:(?-s:.)*|oobar(?-s:.)*)|(?-s:.)*buzz
+		{"f.*|foobar.*|.*buzz", true, newOrFilter(newOrFilter(literalFilter([]byte("f")), literalFilter([]byte("foobar"))), literalFilter([]byte("buzz"))), true},
+		// parsed as ((f(?-s:.)*)|foobar(?-s:.)*)|(?-s:.)*buzz
+		{"((f.*)|foobar.*)|.*buzz", true, newOrFilter(newOrFilter(literalFilter([]byte("f")), literalFilter([]byte("foobar"))), literalFilter([]byte("buzz"))), true},
+
+		// regex we are not supporting.
```

Comment:
Looking forward to see support for these in the future, many of them would be trivial to add.

Reply:
Yes but I wanted to keep the PR small. I think we could ultimately rewrite all regexp, but not with just literal filter.

Reply:
Mainly because we actually filter and not match line.

Reply:
I understand and appreciate that you want to keep PR small. I think it's a right choice. We can always add more cases in future PRs!